lib/iris/fileformats/_nc_load_rules/actions.py

```diff
@@ -367,9 +375,34 @@ def action_build_dimension_coordinate(engine, providescoord_fact):
             assert coord_grid_class in grid_classes
 
     if succeed:
-        hh.build_dimension_coordinate(
+        hh.build_and_add_dimension_coordinate(
             engine, cf_var, coord_name=coord_name, coord_system=coord_system
         )
+
+    else:
+        message = f"Dimension coordinate {var_name} not created. Debug info:\n"
+        if succeed is None:
+            message += "An unexpected error occurred"
+            error = NotImplementedError(message)
+        else:
+            message += rule_name
+            error = ValueError(message)
+
+        try:
+            # TODO: is there a better way to create a stack trace?
+            #  Is an ST even valuable? `error` can be passed straight into
+            #   from_exception() without raising it first.
+            raise error
+        except error.__class__ as error:
+            tb_exception = TracebackException.from_exception(error)
```

Comment:
📢

Reply:
I've tried to follow through the docs on this, and they do say 
[on Traceback objects](https://docs.python.org/3/reference/datamodel.html#traceback-objects) : "A traceback object is implicitly created when an exception occurs, and may also be explicitly created by calling `types.TracebackType`".
But TBH it seems that it is _normally_ the `raise` statement itself tdoes the work of walking the stack and creating a list of `Traceback` objects.  So I think it's still **_far_** easier to just use one to do that, even in this case where you are using a 'raise' yourself for the sake of it, just to produce a stacktrace consistent with other failure mechanisms.

Reply:
After some reading, I'm also in agreement on the use of `TracebackException.from_exception(error)`.
This way, I think we get the equivalent of a stacktrace **_printout_**, but it loses references to actual _objects_ in the frames (like pdb would use).  Which is the right level of information (and will even serialise if needed).

It's possible this still stores too much, if I misunderstood, but in that case I think you could replace it with a printout of the same thing, i.e. a list of strings (like the `.format() you mention).